src/haz3lcore/zipper/ProjectorBase.re

```diff
@@ -0,0 +1,175 @@
+open Util;
+open Virtual_dom.Vdom;
+
+/* This module, along with ZipperBase, exists to resolve
+ * cyclic dependencies between Zipper and Projector. All
+ * projector functionality should be added to Projector.re
+ * or above unless it would create such a cycle */
+
+/* Enumeration of different kinds of projectors. This is
+ * used as a key when adding new projectors to an editor. */
+[@deriving (show({with_path: false}), sexp, yojson)]
+type kind =
+  | Fold
+  | Info
+  | Checkbox
+  | Slider
+  | SliderF
+  | TextArea;
+
+/* The projector map is store alongside the zipper and
+ * maps syntax UUIDs to instantiated projectors. This
+ * is how projector placement and models are persisted */
+[@deriving (show({with_path: false}), sexp, yojson)]
+module Map = {
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type entry = {
+    kind,
+    model: string,
+  };
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type t = Id.Map.t(entry);
+  open Id.Map;
+  let empty = empty;
+  let find = find_opt;
+  let mem = mem;
+  let mapi = mapi;
+  let update = update;
+};
+
+/* Projectors currently have two options for placeholder
+ * shapes: A inline display of a given length, or a block
+ * display with given length & height. Both of these can
+ * depend on the projector model and info package */
+[@deriving (show({with_path: false}), sexp, yojson)]
+type shape =
+  | Inline(int)
+  | Block(Measured.Point.t);
+
+/* The type of syntax which a projector can replace.
+ * Right now projectors can replace a single piece */
+[@deriving (show({with_path: false}), sexp, yojson)]
+type syntax = Piece.t;
+
+/* Global actions available to handlers in all projectors */
+type external_action =
+  | Remove /* Remove projector entirely */
+  | Escape(Util.Direction.t) /* Pass focus to parent editor */
+  | SetSyntax(syntax); /* Set underlying syntax */
+
+/* External info fed to all projectors. Eventually
+ * dynamic information will be added here. Projector
+ * position and dimensions in base editor could be
+ * added here if needed */
+[@deriving (show({with_path: false}), sexp, yojson)]
+type info = {
+  id: Id.t,
+  syntax,
+  ci: option(Info.t),
+};
+
+let info_init = (p: syntax) => {id: Piece.id(p), syntax: p, ci: None};
+
+/* To add a new projector, implement this module signature */
+module type Projector = {
+  [@deriving (show({with_path: false}), sexp, yojson)]
+  type model;
+  /* The internal model type of the projector which will
+   * be serialized and persisted. Use `unit` if you don't
+   * need other state beyond the underlying syntax */
```

Comment:
Not trying to be nitpicky but can we commit to docstrings either above or below defns?